### ISSUE TITLE
Evaluate using Profile-Guided Optimization (PGO) and LLVM BOLT

### ISSUE THREAD

**Issue by zamazan4ik** [defold/defold#8193]
_Is it a private platform specific feature? (i.e. would it break your NDA?)<br>
No.

**Is your feature request related to a problem? Please describe (REQUIRED):**
Not a problem. Just an idea of how to possibly improve the Defold engine performance from the CPU perspective.

I am collecting all materials (benchmarks, articles, stories, showcases, etc.) about Profile-Guided Optimization (PGO) across different applications [here](https://github.com/zamazan4ik/awesome-pgo). I am sure you will be able to find helpful information regarding PGO.

Related to the gamedev domains, I suggest you look at the following things:

* Unreal Engine supports PGO build since 4.27 ([release notes](https://docs.unrealengine.com/4.27/en-US/WhatsNew/Builds/ReleaseNotes/4_27/)). According to the documentation, PGO allows to achieve better performance on UE too (+10% on some CPU-heavy scenarios from this page). Also, I've talked with developers in a local Telegram chat about UE. A person said that they use PGO as a default optimization with UE and their games. The PGO profiles are collected via crafted local test workloads (usually - the most difficult scenes) with [Gauntlet](https://docs.unrealengine.com/4.27/en-US/TestingAndOptimization/Automation/Gauntlet/). The performance improvement is something like 6-8%
* I did some benchmarks for Bevy with PGO. PGO-run (first) vs non-PGO (second) - [Pastebin](https://gist.github.com/zamazan4ik/bbffbdf9b10e2a281f5d5373347f48ef). In these results you need to interpret performance decrease as "Release version is slower than PGOed" and performance increase as "Release version is faster than PGOed". As you see, in many scenarios PGO improves performance, but in some of them, performance is decreased.
* Godot proposal about PGO - [link](https://github.com/godotengine/godot-proposals/issues/2610)
* Unity Burst thread about PGO - [link](https://forum.unity.com/threads/profile-guided-optimisation-for-burst.931419/)

That's why I think trying to optimize Defold with PGO would be a good idea.

**Describe the solution you'd like (REQUIRED):**
I can propose the following action points:

* Check the PGO effects on Defold. If it works well - write a note in the Defold documentation about building with PGO. 
* Provide an easy way to build Defold with PGO (e.g. via custom build options in the build scripts), and experienced users will be able to rebuild Defold with PGO on their own for their own usage scenarios.
* Optimizing prebuilt binaries with PGO would be a nice thing to have (if it's possible to prepare a generic enough workload).

As an additional optimization, I suggest looking at [LLVM BOLT](https://github.com/llvm/llvm-project/blob/main/bolt/README.md). But from my experience, starting with PGO and only then using BOLT would be better.

**Describe alternatives you've considered (REQUIRED):**
Do not perform additional optimization.

**Comment by JCash** [comment]
Hi!

While the PGO feels useful, I also get flashes of shader compiler permutations.

I think it might be doable to add a compile flag, to research any potential gains.

I don't know enough about how the workflow though.
I'm guessing this PGO is only used occasionally? As it otherwise would be subject of a loop of `compile->run test->compile->run test...` which would  take an undue time for every change we do to the code.
How many runs are recommended? 1, 2, 5?

Currently, building the entire engine takes ~50seconds, while running the tests takes ~5 minutes
Rebuilding a library+engine takes a couple of seconds, and adding running all the tests (if only for that library), would be a no go for development.

Is the goal to identify possibly optimizations that the compiler can suggest to us, or is it to "just" optimize the compiled code?

In general, I want the default build (for the devs) to be the same code as the end user runs.
I feel doing this PGO step is a new beast entirely, that we'd need to think more about for our production code.
As a separate parameter though, for the professional ones that want to build their own builds from scratch, it shouldn't be a problem.

I think LLVM BOLT suffers a bit from the same problem.
For us to run the same code as the users, we would have to continually run all the tests first, so we debug the same code as they run.

So, in summary:
* Yes, I think it might be interesting to add a build flag for research
* No, I don't think it should be default behavior at this time

**Comment by zamazan4ik** [comment]
> I'm guessing this PGO is only used occasionally?

Generally speaking, yes, you are right - PGO-optimized builds could be done only for the Releases (when we want to deliver the most optimized version). Definitely, there is no reason to perform PGO optimization for usual development builds.

However, it's possible to precompute training data (`.profraw`/`.gcda` files for Clang/GCC correspondingly, don't know how they are called in MSVC), commit it to the repo, and reuse it for each compilation (or at least when you run benchmarks in CI). It won't add any compile-time overhead because the most time-consuming operation in PGO is training data collection. But in this case, you need to keep in mind the "profile skew" problem. It's the problem when your precomputed training data are skewed from the actual source code. It's not a big problem for rarely-touched source code (because there will be no skew there, obviously) but it could be the case for the fast-evolving codebase parts. Anyway, it could be easily mitigated by regenerating & committing the training data each N commits (N is up to you and could be found empirically or based on code coverage decision - more info is [here](https://github.com/Kobzol/cargo-pgo/issues/30) (the same approach is valid for C++ codebase too)).

> Rebuilding a library+engine takes a couple of seconds, and adding running all the tests (if only for that library), would be a no go for development.

I want to highlight that using the "usual" tests like unit tests is not the best idea for getting PGO results. Because unit-tests often perform testing of many corner cases but with PGO we want to optimize the hot path (and corner cases are not in the hot path by the definition). So perhaps would be a better idea to craft more specific near real-life scenarios and get the PGO profiles from it. But if you already have such tests (I trust you since I have no insights about the current tests infra in Defold) - it's completely ok to use the existing tests.

> How many runs are recommended? 1, 2, 5?

The aim here is to cover with profiles as much code as possible. So if you have various real-life scenarios, would be best if you run all of them. However, each scenario is enough to run only 1 time (if all used for PGO scenarios have equal importance for you).

> Is the goal to identify possibly optimizations that the compiler can suggest to us, or is it to "just" optimize the compiled code?

Hmmm, I am not sure what you meant here :) I suggest at the beginning trying to recompile Defold with PGO and try to measure the possible performance benefits in CPU-bound applications (and in not CPU-bound - measure CPU usage reduction). And if PGO gives something - think about how PGO could be integrated into the Defold build pipelines (e.g. an additional build flag in the build scripts).

> In general, I want the default build (for the devs) to be the same code as the end user runs.
I feel doing this PGO step is a new beast entirely, that we'd need to think more about for our production code.
As a separate parameter though, for the professional ones that want to build their own builds from scratch, it shouldn't be a problem.

Completely agree. PGO is definitely an advanced build option that requires some skills from the end-users to implement it for their own use cases. You as a Defold developer just can try to lower the entrance level to such an optimization for them in the case of Defold. Like UE chose this way (PGO is available via a build option and AFAIK default UE prebuilt binaries are not PGO-optimized yet).

> I think LLVM BOLT suffers a bit from the same problem.
For us to run the same code as the users, we would have to continually run all the tests first, so we debug the same code as they run.

I would suggest not thinking a lot about LLVM BOLT before testing/implementing the PGO approach since PGO should provide more value with less effort, is a more "stable" optimization right now, has fewer limitations (like works on all major platforms (LLVM BOLT works fine only on Linux and experimentally on macOS)).

> So, in summary:

Totally agree with your conclusion!

**Comment by JCash** [comment]
"I want to highlight that using the "usual" tests like unit tests is not the best idea for getting PGO results. Because unit-tests often perform testing of many corner cases but with PGO we want to optimize the hot path "

Right. So for us, we'd have to have a project that triggers _all_ code paths at the same time. I think we'd have a hard time maintaining such a project. We could theoretically synthesize such a project, but that's not without burden either.
Possibly, we could use one of our community games, but question is, which one. ANd should it be 2D or 3D. And how often can we get updates from it? Sounds a bit cumbersome imho.

But, still, if the developer can supply such a .gprof file (or similar), I guess it would be fairly straightforward to pass it along to the build system if they want to build from scratch.

For our build server, given that this is a file given to the linker step, I think it's more feasible.
E.g. if the user can supply a profile file, we can pass it along to the build server, and use that.

I think that's the best of the two worlds: We don't have to do it for our libraries, and the users can optimize their own games individually.
